### PR TITLE
504/adr_IAsyncEnumerable_IActionResult

### PR DESCRIPTION
**Pull request by juzuluag**
Issue Number: (504)

## Summary
ADR 0005 to document the decision.

## Changes

- Add document explaining the decision.

## Checklist

- [ ] Local Tests Passing?
- [ ] CICD and Pipeline Tests Passing?
- [ ] Added any new Tests?
- [x] Documentation Updates Made?
- [ ] Are there any API Changes? If yes, please describe below.
- [ ] This is not a breaking change. If it is, please describe it below.

## Are there API Changes?
N/A
## Is this a breaking change?
N/A

## Anything else?
N/A

### CODE REVIEW

**Comment by bderusha** [comment]
@juzuluag Can you base this branch off of the latest `dev` branch?

**Comment by bderusha** [comment]
@juzuluag Can I make one more request, that we also drop all of the bonus commits from this PR?

File: docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# 5. Using IAsyncEnumerable partial failure won't be handled by HttpResponseExceptionFilter
```
Comment (bderusha) [inline docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md]: Let's reword this as a decision to make:

```markdown
# 5. WebAPI to use IAsyncEnumerable to stream potentially large responses

## Status

Denied
```

OR the positive version::
 
```markdown
# 5. WebAPI to use IActionResult for potentially large resonses.

## Status

Accepted
```

File: docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+## Status
+Accepted
+
+## Context
+There are currently two endpoints (`emissions/forecasts/batch` , `emissions/average-carbon-intensity/batch`) that require passing a payload as an array of items. This array can be have as many items as the user wants, and it can take sometime to be processed which incurs in a delay to the client, which might feel that the service is irresponsive. Also, since these requests return enumerable items, and those are buffered before the client gets it, it impacts the overall memory footprint of the WebApp. By changing the signature of these endpoints to return an `IAsyncEnumerable` collection, it would help to deal with these issues.
```
Comment (bderusha) [inline docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md]:
```suggestion
There are currently two endpoints (`emissions/forecasts/batch` , `emissions/average-carbon-intensity/batch`) that require passing a payload as an array of items. This array can be have as many items as the user wants, and it can take sometime to be processed which incurs in a delay to the client, which might feel that the service is irresponsive. Also, since these requests return enumerable items, and those are buffered before the client gets it, it impacts the overall memory footprint of the WebApp. Changing the signature of these endpoints to return an `IAsyncEnumerable` collection streams the response and helps to deal with these memory concerns.
```

File: docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+## Decision
```
Comment (bderusha) [inline docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md:9]: Depending on which approach you take, `Use IAsyncEnumerable: Denied` or `Use IActionResult: Accepted`, can we put the relevant pros into the  `## Decision`, and the cons into the `## Consequences` sections.

File: docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md (hunk 1 above)
Comment (juzuluag) [reply docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md]: Good. I'll add the former.

**Review by bderusha — APPROVED** [review]

File: docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# 5. WebAPI to use IAsyncEnumerable to stream potentially large responses.
+
+## Status
+Denied
+
+## Context
+There are currently two endpoints (`emissions/forecasts/batch`, `emissions/average-carbon-intensity/batch`) that require passing a payload as an array of items. This array can have as many items as the user wants, and it can take sometime to be processed which incurs in a delay to the client, which might feel that the service is irresponsive. Also, since these requests return enumerable items, and those are buffered before the client gets it, it impacts the overall memory footprint of the WebApp. Changing the signature of these endpoints to return an `IAsyncEnumerable` collection streams the response and helps to deal with these memory concerns.
```
Comment (esgraham) [inline docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md]: Nit:  'sometime' should be 'some time' and 'incurs in' should be 'incurs' (or even better would be the word 'creates').

**Review by esgraham — APPROVED** [review]
